File: mRemoteNG/Connection/Connection.Icon.vb
Imports System.ComponentModel
Imports mRemoteNG.App.Runtime

Namespace Connection
    Public Class Icon
        Inherits StringConverter

        Public Shared Icons As String() = New String() {}

        Public Overloads Overrides Function GetStandardValues(ByVal context As System.ComponentModel.ITypeDescriptorContext) As System.ComponentModel.TypeConverter.StandardValuesCollection
            Return New StandardValuesCollection(Icons)
        End Function

        Public Overloads Overrides Function GetStandardValuesExclusive(ByVal context As System.ComponentModel.ITypeDescriptorContext) As Boolean
            Return True
        End Function

        Public Overloads Overrides Function GetStandardValuesSupported(ByVal context As ITypeDescriptorContext) As Boolean
            Return True
        End Function

        Public Shared Function FromString(ByVal IconName As String) As Drawing.Icon
            Try
                Dim IconPath As String = My.Application.Info.DirectoryPath & "\Icons\" & IconName & ".ico"

                If IO.File.Exists(IconPath) Then
                    Dim nI As New Drawing.Icon(IconPath)

                    Return nI
                End If
            Catch ex As Exception
                MessageCollector.AddMessage(Messages.MessageClass.ErrorMsg, "Couldn't get Icon from String" & vbNewLine & ex.Message)
            End Try

            Return Nothing
        End Function
    End Class
End Namespace

File: mRemoteNG/Connection/ConnectionIcon.cs
using System;
using System.ComponentModel;
using System.Runtime.Versioning;
using mRemoteNG.App;
using mRemoteNG.App.Info;


namespace mRemoteNG.Connection
{
    [SupportedOSPlatform("windows")]
    public class ConnectionIcon : StringConverter
    {
        public static string[] Icons = { };

        public override StandardValuesCollection GetStandardValues(ITypeDescriptorContext context)
        {
            return new StandardValuesCollection(Icons);
        }

        public override bool GetStandardValuesExclusive(ITypeDescriptorContext context)
        {
            return true;
        }

        public override bool GetStandardValuesSupported(ITypeDescriptorContext context)
        {
            return true;
        }

        public static System.Drawing.Icon FromString(string iconName)
        {
            try
            {
                var iconPath = $"{GeneralAppInfo.HomePath}\\Icons\\{iconName}.ico";

                if (System.IO.File.Exists(iconPath))
                {
                    var nI = new System.Drawing.Icon(iconPath);
                    return nI;
                }
            }
            catch (Exception ex)
            {
                Runtime.MessageCollector.AddMessage(Messages.MessageClass.ErrorMsg,
                                                    $"Couldn\'t get Icon from String" + Environment.NewLine +
                                                    ex.Message);
            }

            return null;
        }
    }
}